src/daemon/messages.go

```diff
@@ -230,54 +230,73 @@ type IntroductionMessage struct {
 	Port uint16
 	// Our client version
 	Version int32
-
-	c *gnet.MessageContext `enc:"-"`
+	c       *gnet.MessageContext `enc:"-"`
 	// We validate the message in Handle() and cache the result for Process()
 	valid bool `enc:"-"` // skip it during encoding
+	// Extra would be parsed as blockchain pubkey if it's not empty
+	Extra []byte `enc:",omitempty"`
 }
 
 // NewIntroductionMessage creates introduction message
-func NewIntroductionMessage(mirror uint32, version int32, port uint16) *IntroductionMessage {
+func NewIntroductionMessage(mirror uint32, version int32, port uint16, extra []byte) *IntroductionMessage {
 	return &IntroductionMessage{
 		Mirror:  mirror,
 		Version: version,
 		Port:    port,
+		Extra:   extra,
 	}
 }
 
 // Handle Responds to an gnet.Pool event. We implement Handle() here because we
 // need to control the DisconnectReason sent back to gnet.  We still implement
 // Process(), where we do modifications that are not threadsafe
 func (intro *IntroductionMessage) Handle(mc *gnet.MessageContext, daemon interface{}) error {
-	d := daemon.(*Daemon)
+	d := daemon.(Daemoner)
 
 	err := func() error {
 		// Disconnect if this is a self connection (we have the same mirror value)
-		if intro.Mirror == d.Messages.Mirror {
+		if intro.Mirror == d.Mirror() {
 			logger.Infof("Remote mirror value %v matches ours", intro.Mirror)
-			d.Pool.Pool.Disconnect(mc.Addr, ErrDisconnectSelf)
+			d.Disconnect(mc.Addr, ErrDisconnectSelf)
 			return ErrDisconnectSelf
 
 		}
 
 		// Disconnect if not running the same version
-		if intro.Version != d.Config.Version {
+		if intro.Version != d.DaemonConfig().Version {
 			logger.Infof("%s has different version %d. Disconnecting.",
 				mc.Addr, intro.Version)
-			d.Pool.Pool.Disconnect(mc.Addr, ErrDisconnectInvalidVersion)
+			d.Disconnect(mc.Addr, ErrDisconnectInvalidVersion)
 			return ErrDisconnectInvalidVersion
 		}
 
 		logger.Infof("%s verified for version %d", mc.Addr, intro.Version)
 
+		// Checks the genesis hash if not empty
+		if len(intro.Extra) > 0 {
+			var bcPubKey cipher.PubKey
+			if len(intro.Extra) != len(bcPubKey) {
```

Comment:
I think we want to ignore the length check for the ExtraData.  Otherwise, if we want to add another field later, we have to redo the 3-version upgrade again.  The length should meet the minimum, but if it is too long, we can ignore the extra data.